From 149c98aeaad0a6022543d9e3b36617bb1ac006f1 Mon Sep 17 00:00:00 2001
From: Lennart Poettering <lennart@poettering.net>
Date: Mon, 18 Aug 2014 23:15:51 +0200
Subject: [PATCH] util: try to be a bit more NFS compatible when checking
 whether an FS is writable

https://bugs.freedesktop.org/show_bug.cgi?id=81169
(cherry picked from commit 70421bdce2719d76efffd8afdc28433c75aac5a2)
---
 src/shared/path-util.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/src/shared/path-util.c b/src/shared/path-util.c
index 5c0bf93181..55a8e61fec 100644
--- a/src/shared/path-util.c
+++ b/src/shared/path-util.c
@@ -446,7 +446,16 @@ int path_is_read_only_fs(const char *path) {
         if (statvfs(path, &st) < 0)
                 return -errno;
 
-        return !!(st.f_flag & ST_RDONLY);
+        if (st.f_flag & ST_RDONLY)
+                return true;
+
+        /* On NFS, statvfs() might not reflect whether we can actually
+         * write to the remote share. Let's try again with
+         * access(W_OK) which is more reliable, at least sometimes. */
+        if (access(path, W_OK) < 0 && errno == EROFS)
+                return true;
+
+        return false;
 }
 
 int path_is_os_tree(const char *path) {
